### DISCUSSION TITLE
Can't scroll to id when listening to router.events and setting loading animation

### DISCUSSION THREAD

**Discussion by DiegoMcDipster** [vercel/next.js#57510] [Help]
### Summary

I want to be able to scroll from page1 to the contact section on my homepage.  However, my application listens for router events and kicks off a loading animation.  The animation seems to throw off the scrolling to section. (note, to keep the example simple, I've replaced the animation with the text '...loading')

For example, I click the link from page1 to #contact.  
Scenario 1: with loading animation on the _app.tsx: 
```
return <>{isLoading ? <p>...loading</p> : <Component {...pageProps} />}</>;
```
It directs me to the top of the homepage.

Scenario 2: with animation removed from returned jsx
```
return <Component {...pageProps} />;
```
It will scroll to the contact section.  

How can I get around this? So that the loading animation does run AND it scrolls to the contact section?

### Additional information

```js
Here's my _app.tsx:

export default function App({ Component, pageProps }: AppProps) {
  const router = useRouter();
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const handleRouteChangeStart = () => {
      setIsLoading(true);
    };

    const handleRouteChangeComplete = () => {
      setIsLoading(false);
    };

    const handleRouteChangeError = () => {
      setIsLoading(false);
    };

    router.events.on("routeChangeStart", handleRouteChangeStart);
    router.events.on("routeChangeComplete", handleRouteChangeComplete);
    router.events.on("routeChangeError", handleRouteChangeError);

    return () => {
      router.events.off("routeChangeStart", handleRouteChangeStart);
      router.events.off("routeChangeComplete", handleRouteChangeComplete);
      router.events.off("routeChangeError", handleRouteChangeError);
    };
  }, [router.events]);

  return <>{isLoading ? <p>...loading</p> : <Component {...pageProps} />}</>;
}
```

Here's the html from my page1: 
```
<main className={styles.main}>
        <Link style={{ margin: "2rem", fontSize: "4rem" }} href={"/#contact"}>
          Contact
        </Link>
      </main>
```

Here's my homepage (index.tsx):
```
<main className={styles.main}>
        <Link style={{ margin: "2rem", fontSize: "4rem" }} href={"/page1"}>
          Page1
        </Link>
        <br />
        <section className="box box-banner">Banner</section>
        <section className="box box-how">The how</section>
        <section className="box box-benefits">The benefits</section>
        <section id="contact" className="box box-contact">
          Contact
        </section>
      </main>
```

And the css:
```
.box {
  border: 2px solid gold;
  width: 300px;
  height: 800px;
}

.box-banner {
  background-color: dodgerblue;
}

.box-how {
  background-color: tomato;
}

.box-benefits {
  background-color: gold;
}

.box-contact {
  background-color: green;
}
```
```


### Example

_No response_

**Comment by icyJoseph** [comment]
Right, so the issue is that when you set the loading animation, you tell React to unmount your application, so the target element that you want to scroll to gets destroyed. You need to make sure that you persist the Page (Component) 

Try something like:

```tsx
  return (
    <>
      <p className={`loader ${isLoading ? "loading" : "idle"}`}>Loading...</p>
      <Component {...pageProps} />
    </>
  );
```

You can also remove the `loader`, p tag in this case, but it might be acceptable, perhaps wanted, to keep it mounted.

**Comment by DiegoMcDipster** [comment]
@icyJoseph Thanks for taking the time to reply.

So, from what i've understood I can use the loading flag to set the CSS of the loading animation as follows:
```
.loading {
  display: block;
}

.idle {
  display: none;
}
```

That does do the trick on my test env. However, thinking about my actual app, which does data fetching in gssp on some route changes, doesn't this mean that the page will always be visible just under the loading animation during all route changes?

**Reply by icyJoseph** [reply]
Ah, I guess so, you can also put the loader after the Component rendering.

You can also unmount the loader and alsp use hash router changes instead, if you only want those to trigger the animation.

- https://nextjs.org/docs/pages/api-reference/functions/use-router#routerevents

**Comment by DiegoMcDipster** [comment]
In the end I gave up on it.  the solutions seem too hacky for me. I seperated Contact out onto another page. And that works fine.

Thanks for giving me those suggestions!
